pkg/kubeapiserver/options/admission.go

```diff
@@ -0,0 +1,123 @@
+//  all kube-apiserver admission plugins.
+//
+//  Provides the list of RecommendedPluginOrder that holds sane values
+//  that can be used by servers that don't care about admission chain.
+//  Servers that do care can overwrite/append that field after creation.
+func NewAdmissionOptions() *AdmissionOptions {
+	options := genericoptions.NewAdmissionOptions()
+	// register all admission plugins
+	RegisterAllAdmissionPlugins(options.Plugins)
+	// set RecommendedPluginOrder
+	options.RecommendedPluginOrder = AllOrderedPlugins
+	// set DefaultOffPlugins
+	options.DefaultOffPlugins = DefaultOffAdmissionPlugins()
+
+	return &AdmissionOptions{
+		GenericAdmission: options,
+	}
+}
+
+// AddFlags adds flags related to admission for kube-apiserver to the specified FlagSet
+func (a *AdmissionOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.StringSliceVar(&a.PluginNames, "admission-control", a.PluginNames, ""+
+		"Admission is divided into two phases. "+
+		"In the first phase, only mutating admission plugins run. "+
+		"In the second phase, only validating admission plugins run. "+
+		"The names in the below list may represent a validating plugin, a mutating plugin, or both. "+
+		"The order of plugins in which they are passed to this flag does not matter. "+
+		"Comma-delimited list of: "+strings.Join(a.GenericAdmission.Plugins.Registered(), ", ")+".")
+	fs.MarkDeprecated("admission-control", "Use --enable-admission-plugins or --disable-admission-plugins instead. Will be removed in a future version.")
+
+	a.GenericAdmission.AddFlags(fs)
+}
+
+// Validate verifies flags passed to kube-apiserver AdmissionOptions.
+// Kube-apiserver verifies PluginNames and then call generic AdmissionOptions.Validate.
+func (a *AdmissionOptions) Validate() []error {
+	if a == nil {
+		return nil
+	}
+	errs := []error{}
+	if a.PluginNames != nil &&
+		(a.GenericAdmission.EnablePlugins != nil || a.GenericAdmission.DisablePlugins != nil) {
+		errs = append(errs, fmt.Errorf("admission-control and enable-admission-plugins/disable-admission-plugins flags are mutually exclusive"))
+	}
+
+	registeredPlugins := sets.NewString(a.GenericAdmission.Plugins.Registered()...)
+	for _, name := range a.PluginNames {
+		if !registeredPlugins.Has(name) {
+			errs = append(errs, fmt.Errorf("admission-control plugin %q is unknown", name))
+		}
+	}
+
+	errs = append(errs, a.GenericAdmission.Validate()...)
+
+	return errs
+}
+
+// ApplyTo adds the admission chain to the server configuration.
+// Kube-apiserver just call generic AdmissionOptions.ApplyTo.
+func (a *AdmissionOptions) ApplyTo(
+	c *server.Config,
+	informers informers.SharedInformerFactory,
+	kubeAPIServerClientConfig *rest.Config,
+	scheme *runtime.Scheme,
+	pluginInitializers ...admission.PluginInitializer,
+) error {
+	if a == nil {
+		return nil
+	}
+
+	if a.PluginNames != nil {
+		// pass PluginNames to generic AdmissionOptions
+		a.GenericAdmission.EnablePlugins = a.PluginNames
```

Comment:
You need to calculate the disables as well. Otherwise a default enabled plugin will remain enabled, but compatibility demands that it be disabled.

Reply:
> deads2k just now  Member
> You need to calculate the disables as well. Otherwise a default enabled plugin will remain enabled, but compatibility demands that it be disabled.

and add a test that validates that.  I glossed over it in my last review.

Reply:
DefaultOffPlugins is set in NewAdmissionOptions, and also processed in genericapiserver.

Reply:
And sure we can disable a default enabled plugin.

Reply:
>and add a test that validates that. 

Yes, there is already a test case https://github.com/hzxuzhonghu/kubernetes/blob/17c4045bc15a346e33775b997710d64718abf8a1/staging/src/k8s.io/apiserver/pkg/server/options/admission_test.go#L63~L68

Reply:
> Yes, there is already a test case https://github.com/hzxuzhonghu/kubernetes/blob/17c4045bc15a346e33775b997710d64718abf8a1/staging/src/k8s.io/apiserver/pkg/server/options/admission_test.go#L63~L68

It looks like that test is focused on the normal calculation.  What I'm pointing how here is that if a user specifices `--admission-control=one, two` for kube-apiserver and recommended is "one, two, three, four" and defaultOff is "four", the user expects to end up with "one, two", but actually gets "one, two, three", right?

Reply:
Get it, I will add this case.

Reply:
@deads2k Done https://github.com/hzxuzhonghu/kubernetes/blob/dd9ba4d84f5d74942302c53b8b465bb367c37a8e/staging/src/k8s.io/apiserver/pkg/server/options/admission_test.go#L81-L95